EKS SQL MI scenario (#523)
microsoft/azure_arc: docs/azure_arc_jumpstart/azure_arc_data/eks/eks_mssql_mi/_index.md

---
type: docs
title: "Data Controller Terraform plan"

> linkTitle: "Data Controller Terraform plan"

Reviewer: It's not Vanilla

weight: 1
description: >
---

> ## Deploy an Azure Arc Data Controller (Vanilla) on EKS using Terraform

Reviewer: It's not Vanilla

* EKS cleanup
* *terraform destroy*

## Prerequisites

* Clone the Azure Arc Jumpstart repository

  ```shell
  git clone https://github.com/microsoft/azure_arc.git
  ```

* [Install AWS IAM Authenticator](https://docs.aws.amazon.com/eks/latest/userguide/install-aws-iam-authenticator.html)

* [Install or update Azure CLI to version 2.15.0 and above](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). Use the below command to check your current installed version.

  ```shell
  az --version
  ```

* [Create a free Amazon Web Services account](https://aws.amazon.com/free/) if you don't already have one.

* [Install Terraform >=0.15](https://learn.hashicorp.com/terraform/getting-started/install.html)

* Create Azure service principal (SP)

  To be able to complete the scenario and its related automation, Azure service principal assigned with the “Contributor” role is required. To create it, login to your Azure account run the below command (this can also be done in [Azure Cloud Shell](https://shell.azure.com/)).

  ```shell
  az login
  az ad sp create-for-rbac -n "<Unique SP Name>" --role contributor
  ```

  For example:

  ```shell

>   az ad sp create-for-rbac -n "http://AzureArcK8s" --role contributor

Reviewer: ```suggestion
  az ad sp create-for-rbac -n "http://AzureArcData" --role contributor
```

  ```

  Output should look like this

  ```json
  {
  "appId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",

>   "displayName": "AzureArcK8s",
>   "name": "http://AzureArcK8s",

Reviewer: ```suggestion
  "displayName": "AzureArcData",
  "name": "http://AzureArcData",
```

* In addition, the plan will copy the EKS *kubeconfig* file as well as the *configmap.yml* file (which is responsible for having the EKS nodes communicate with the cluster control plane) on to the Windows instance.

* As part of the Windows Server 2019 VM deployment, there are 3 scripts executions:

  1. *azure_arc.ps1* script will be created automatically as part of the Terraform plan runtime and is responsible on injecting the *TF_VAR* variables values on to the Windows instance which will then be used in both the *ClientTools* and the *LogonScript* scripts.

  2. *ClientTools.ps1* script will run at the Terraform plan runtime Runtime and will:
      * Create the *ClientTools.log* file  
      * Install the required tools – az cli, az cli Powershell module, kubernetes-cli, aws-iam-authenticator, Visual C++ Redistributable (Chocolaty packages)
      * Download Azure Data Studio & Azure Data CLI

>       * Download the *DC_Cleanup* and *DC_Deploy* Powershell scripts

Reviewer: Update script file names

      * Create the logon script
      * Create the Windows schedule task to run the logon script at first login

  3. *LogonScript.ps1* script will run on user first logon to Windows and will:
      * Create the *LogonScript.log* file
      * Install Azure Data Studio & Azure Data CLI
      * Install the Azure Data Studio Azure Data CLI, Azure Arc & PostgreSQL extensions
      * Apply the *configmap.yml* file on the EKS cluster
      * Create the *azdata* config file in user Windows profile
      * Create the Azure Data Studio desktop shortcut
      * Open another Powershell session which will execute a command to watch the deployed Azure Arc Data Controller Kubernetes pods
      * Deploy the Arc Data Controller using the *TF_VAR* variables values
      * Deploy the MSSQL managed instance onto the data controller
      * Restore a sample AdventureWorks database onto the data controller
      * Unregister the logon script Windows schedule task so it will not run after first login
      * Open Azure Data Studio

## Deployment

> As mentioned, the Terraform plan will deploy an EKS cluster, the Azure Arc Data Controller on that cluster and an EC2 Windows Server 2019 Client instance.

Reviewer: ```suggestion
As mentioned, the Terraform plan will deploy an EKS cluster, the Azure Arc Data Controller on that cluster, a SQL Managed Instance with a sample database, and a Windows Server 2019 Client GCP compute instance.
```

  ![RDP to the Client instance](./24.png)

  ![RDP to the Client instance](./25.png)

  ![RDP to the Client instance](./26.png)

* Using the decrypted password, RDP the Windows instance. In case you need to get the password later, use the ```terraform output``` command to re-present the plan output.

* At first login, as mentioned in the "Automation Flow" section, a logon script will get executed. This script was created as part of the automated deployment process.

* Let the script to run it's course and **do not close** the PowerShell session, this will be done for you once completed. You will notice that the Azure Arc Data Controller gets deployed on the EKS cluster. **The logon script run time is approximately 10min long**.

    Once the script will finish it's run, the logon script PowerShell session will be close and the Azure Arc Data Controller will be deployed on the EKS cluster and be ready to use.

    ![PowerShell login script run](./27.png)

    ![PowerShell login script run](./28.png)

    ![PowerShell logon script run](./29.png)

    ![PowerShell login script run](./30.png)

    ![PowerShell login script run](./31.png)

    ![PowerShell logon script run](./32.png)

>     ![Data Controller in a resource group](./33.png)

Reviewer: If you deployed in indirect mode, this should not be here.

>     ![Data Controller resource](./34.png)

Reviewer: If you deployed in indirect mode, this should not be here.

* Using PowerShell, login to the Data Controller and check it's health using the below commands.

  ```shell
  azdata login --namespace $env:ARC_DC_NAME
  azdata arc dc status show
  ```

  ![azdata login](./35.png)

* Another tool automatically deployed is Azure Data Studio along with the *Azure Data CLI*, the *Azure Arc* and the *PostgreSQL* extensions. Azure Data Studio is automatically opened after the deployment finishes.

* From Azure Data Studio, click on the MSSQL_MI instance and view the sample AdventureWorks database.

  ![Sample AdventureWorks database](./36.png)

## Cleanup

* To delete the Azure Arc Data Controller and all of it's Kubernetes resources, run the *MSSQL_MI_Cleanup.ps1* PowerShell script located in *C:\tmp* on the Windows Client instance. At the end of it's run, the script will close all PowerShell sessions. **The Cleanup script run time is ~2-3min long**.

>   ![MSSQL_MI_Cleanup PowerShell script run](./38.png)

Reviewer: png jump 36<>38
